internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,203 @@
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return batch
+	}
+
+	allMetricNames := buildMetricNames(prefix)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := allMetricNames[name]; !ok {
+			continue
+		}
+
+		batch.Add(validatePromLint(name))
+		batch.Add(validateName(name, prefix))
+
+		labels := map[string]string{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = *label.Value
+			}
+		}
+		batch.Add(validateLabels(name, labels))
+
+		// after we validate the metric, delete it from the set of all
+		// expected metrics so that we can track if any were not found.
+		delete(allMetricNames, name)
+	}
+
+	if len(allMetricNames) > 0 {
+		notFound := make([]string, 0, len(allMetricNames))
+		for k := range allMetricNames {
+			notFound = append(notFound, k)
+		}
+		batch.Add(fmt.Errorf("%w metrics %v", errNotFound, notFound))
+	}
+
+	return batch
+}
+
+// buildMetricNames creates a set of all the expected metrics names for the given prefix
+// that should be registered with Prometheus as defined in the baseplate spec.
+// The prefix will be either thrift, http, or grpc.
+func buildMetricNames(prefix string) map[string]struct{} {
+	suffixes := []string{"_latency_seconds", "_requests_total", "_active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prefix+"_"+server+suffix] = struct{}{}
+		names[prefix+"_"+client+suffix] = struct{}{}
+	}
+	return names
+}
+
+// validateName checks the following:
+// 1) the metric name has at least 3 parts separated by "_".
+// The parts are <namespace>_<client/server>_<name>_<suffix>, where suffix is optional.
+// Ref: https://prometheus.io/docs/practices/naming
+// 2) the metric name has the correct prefix.
+// 3) the metric contains either "client" or "server" as the second part.
+func validateName(name, prefix string) error {
+	var batch errorsbp.Batch
+
+	const (
+		separator = "_"
+		partCount = 3
+	)
+	parts := strings.SplitN(name, separator, partCount)
+	if len(parts) < partCount {
+		batch.Add(fmt.Errorf("%w metric %q does not have minimum number (%d) of parts", errLength, name, partCount))
+		return batch.Compile()
+	}
+
+	if got, want := parts[0], prefix; got != want {
+		batch.Add(fmt.Errorf("%w metric %q does not have the expected prefix %q", errPrefix, name, prefix))
+	}
+
+	if parts[1] != client && parts[1] != server {
+		batch.Add(fmt.Errorf("%w metric %q", errClientServer, name))
+	}
+	return batch.Compile()
+}
+
+func validateLabels(name string, gotLabels map[string]string) error {
+	wantLabels := buildLables(name)
+
+	var batch errorsbp.Batch
+	for wantLabel := range wantLabels {
+		if _, ok := gotLabels[wantLabel]; !ok {
+			batch.Add(fmt.Errorf("%w metric %q label %q", errLabelNotFound, name, wantLabel))
+		}
+		delete(wantLabels, wantLabel)
+	}
+
+	if len(wantLabels) > 0 {
+		// todo: err
+	}
+	return batch.Compile()
+}
```

Comment:
I would use `cmp.Diff`

```
// var errDiffLabels = fmt.Errorf("labels incorrect")
if diff := cmp.Diff(gotLabels, wantLabels); diff != "" {
  return fmt.Errorf("%w: (-got +want)\n%s", errDiffLabels, diff)
}
```

cmp understands maps and lengths and prints out a nice diff that could be like

```diff
 {
-  "grpc_service": true,
+  "grpc_serice": true,
   "grpc_method": true,
   ...
 }
```

Reply:
perfection! thanks.